### PR TITLE
MIAD-267 create empty plan

### PR DESCRIPTION
**Pull request by anthony-britton-moj**

### CODE REVIEW

**Review by iterative-it — DISMISSED** [review]
LGTM. Like that you've gone with storing the prisoner details

File: src/test/resources/application-test.yml [hunk 1]
```diff
@@ -35,6 +35,29 @@ hmpps.sqs:
...
+            "person.alert.deleted",
+            "person.alerts.changed",
+            "prisoner-alerts.alert-code-created",
+            "prisoner-alerts.alert-code-deactivated",
+            "prisoner-alerts.alert-code-reactivated",
+            "prisoner-alerts.alert-code-updated",
+            "prisoner-alerts.alert-type-created",
+            "prisoner-alerts.alert-type-deactivated",
+            "prisoner-alerts.alert-type-reactivated",
+            "prisoner-alerts.alert-type-updated"
```
Comment (iterative-it) [inline src/test/resources/application-test.yml:53]: Are these needed to continue to support the integration tests?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/service/event/DomainEventsListener.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+@Service
+class DomainEventsListener(
+  private val objectMapper: ObjectMapper,
+  private val personUpdatedHandler: PersonUpdatedHandler,
+) {
+  @SqsListener("hmppsdomaineventsqueue", factory = "hmppsQueueContainerFactoryProxy")
+  fun receive(notification: Notification) {
+    when (notification.eventType) {
+      PRISONER_UPDATED -> personUpdatedHandler.handle(objectMapper.readValue(notification.message))
```
Comment (iterative-it) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/service/event/DomainEventsListener.kt:16]: I imagine we'll also need to listen for merge events and delete the merge from prison number record so we don't accidentally add an alert for a removed number

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/BulkAlertsController.kt [hunk 3]
```diff
@@ -1,44 +1,39 @@
...
-) {
-  @ResponseStatus(HttpStatus.CREATED)
-  @PostMapping
-  @Operation(
-    summary = "Create alerts for multiple people in bulk",
-  )
+class BulkAlertsController(private val plan: PlanBulkAlert) {
+
+  @PostMapping("/plan")
+  @Operation(summary = "Create the plan for bulk alerts")
```
Comment (bill-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/resource/BulkAlertsController.kt:30]: i expected it to be a separate new endpoint, and the old endpoint to be deprecated later.
but Joe HD cfm'd that it's ok for bulk alerts to stop working for dev today, so replacing the old endpoint is even better!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/service/PlanBulkAlert.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import uk.gov.justice.digital.hmpps.hmppsalertsapi.entity.BulkPlan as Plan
+
+@Service
+class PlanBulkAlert(private val planRepository: BulkPlanRepository) {
+  fun createNew(): BulkPlan = planRepository.save(Plan()).toModel()
+  fun update(id: UUID, request: BulkRequest): BulkPlan {
+    val plan = planRepository.getPlan(id)
+
+    return plan.toModel()
+  }
```
Comment (bill-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/service/PlanBulkAlert.kt:18]: i see that this is a placeholder at the moment?
the UI won't need a response body for this.
anyway, as long as it's not returning the full list of added prisoners in the response, i see no problem in returning a response body.

**Review by bill-moj — DISMISSED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/service/PlanBulkAlert.kt [hunk 5]
```diff
@@ -1,21 +1,77 @@
...
     return plan.toModel()
   }
+
+  private fun Plan.setAlertCode(action: SetAlertCode) {
+    alertCode = alertCodeRepository.getByCode(action.alertCode)
+  }
+
+  private fun Plan.setDescription(action: SetDescription) {
+    description = action.description
+  }
```
Comment (bill-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/service/PlanBulkAlert.kt:49]: UI might need to set description to null.

eg, user starts with CSIP alertCode and provides a description text.
these are submitted to backend to get estimated number of alerts to create/update.

then user might go back to change alertCode to DOCGM, and unset description text.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/entity/BulkPlan.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  val createdBy: String = get().username
+  val createdByDisplayName: String = get().userDisplayName
+
+  @ManyToMany
+  @JoinTable(
+    name = "plan_person",
+    joinColumns = [JoinColumn(name = "plan_id")],
+    inverseJoinColumns = [JoinColumn(name = "prison_number")],
+  )
+  val people: MutableSet<PersonSummary> = mutableSetOf()
```
Comment (bill-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsalertsapi/entity/BulkPlan.kt:43]: does it lazy fetch by default?
we don't need to query 10k person summaries when we are just updating the description text.

and i think we need `cleanupMode` for this entity.

**Review by bill-moj — APPROVED** [review]

**Review by iterative-it — APPROVED** [review]
LGTM

**Pull request merged**